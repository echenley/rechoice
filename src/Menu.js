// @flow

import * as React from 'react'
import cx from 'classnames'
import DefaultMenu from './DefaultMenu'
import type {
  ClassMap,
  Item as ItemType,
  ItemRenderer,
  MenuRenderer,
  LoadingState,
} from './types'

type Props = {
  classes: ClassMap,
  controlsId: string,
  focusedItemIndex: ?number,
  getItemId: (i: number) => string,
  inputValue: string,
  isItemSelected: (item: ItemType) => boolean,
  isOpen: boolean,
  itemKey: string,
  items: Array<ItemType>,
  loadingIndicator?: React.Node,
  onHoverItem: (i: number) => mixed,
  onSelect: (item: ItemType) => mixed,
  renderItem: ItemRenderer,
  renderMenu?: MenuRenderer,
  focusedItemRef: (ref: ?HTMLElement) => mixed,
  menuRef: (ref: ?HTMLElement) => mixed,
  state?: LoadingState,
}

class Menu extends React.Component<Props> {
  renderItems = () => {
    const {
      classes,
      focusedItemIndex,
      getItemId,
      inputValue,
      isItemSelected,
      itemKey,
      items,
      onHoverItem,
      onSelect,
      renderItem,
      focusedItemRef,
    } = this.props

    return items.map((item, i) => {
      const isSelected = isItemSelected(item)
      const isFocused =
        typeof focusedItemIndex === 'number' && i === focusedItemIndex

      const renderedItem = renderItem({
        inputValue,
        isFocused,
        isSelected,
        item,
      })

      const itemClasses = cx(classes.item, renderedItem.props.className, {
        [classes.focused]: isFocused,
        [classes.selected]: isSelected,
      })

      return React.cloneElement(renderedItem, {
        'aria-selected': isFocused,
        className: itemClasses,
        id: getItemId(i),
        key: item[itemKey],
        onClick: e => {
          if (renderedItem.props.onClick) {
            renderedItem.props.onClick(e)
          }

          onSelect(item)
        },
        onMouseEnter: () => onHoverItem(i),
        ref: isFocused ? focusedItemRef : null,
        role: 'option',
      })
    })
  }

  render() {
    const {
      classes,
      controlsId,
      inputValue,
      isOpen,
      loadingIndicator,
      renderMenu = DefaultMenu,
      menuRef,
      state = 'success',
    } = this.props

    const menu = renderMenu({
      children: this.renderItems(),
      className: classes.menu,
      listProps: {
        id: controlsId,
        role: 'listbox',
      },
      inputValue,
      isOpen,
      loadingIndicator,
      menuRef,
      state,
    })

    if (!menu) {
      return menu
    }

    if (menu.type === 'string') {
      return React.cloneElement(React.Children.only(menu), {
        ref: menuRef,
      })
    }

    return menu
  }
}

export default Menu
