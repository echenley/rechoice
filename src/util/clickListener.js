// @flow

export function addClickListener(cb: EventHandler) {
  document.addEventListener('mousedown', cb)
  document.addEventListener('touchstart', cb)
}

export function removeClickListener(cb: EventHandler) {
  document.removeEventListener('mousedown', cb)
  document.removeEventListener('touchstart', cb)
}
