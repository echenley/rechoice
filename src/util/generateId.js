// @flow

let i = 0

export default function generateId(prefix: string) {
  return `${prefix}-${i++}`
}
