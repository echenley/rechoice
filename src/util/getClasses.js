// @flow

import type { ClassMap } from '../types'

export default (classes: $Shape<ClassMap>) => ({
  disabled: 'disabled',
  focused: 'focused',
  input: 'rechoice-input',
  inputMirror: 'rechoice-input-mirror',
  inputWrapper: 'rechoice-input-wrapper',
  item: 'rechoice-item',
  menu: 'rechoice-menu',
  root: 'rechoice',
  selected: 'selected',
  tags: 'rechoice-tags',
  value: 'rechoice-value',
  ...classes,
})
