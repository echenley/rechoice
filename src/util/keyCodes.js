// @flow

export default {
  BACKSPACE: 8,
  DELETE: 46,
  ENTER: 13,
  ESCAPE: 27,
  UP: 38,
  DOWN: 40,
  SPACEBAR: 32,
}
