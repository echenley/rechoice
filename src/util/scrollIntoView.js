// @flow

type NodeFinder = (node: HTMLElement) => boolean

export function findParent(
  finder: NodeFinder,
  node: ?HTMLElement,
  rootNode: HTMLElement
) {
  if (node && node !== rootNode.parentNode) {
    if (finder(node)) {
      return node
    }

    return findParent(finder, (node.parentNode: ?$Subtype<Node>), rootNode)
  }

  return null
}

const isScrollable = node => node.scrollHeight > node.clientHeight
const getClosestScrollParent = findParent.bind(null, isScrollable)

export default function scrollIntoView(
  node: HTMLElement,
  rootNode: HTMLElement
) {
  const scrollParent = getClosestScrollParent(node, rootNode)

  if (scrollParent === null) {
    return
  }

  const scrollParentStyles = getComputedStyle(scrollParent)
  const scrollParentRect = scrollParent.getBoundingClientRect()
  const scrollParentBorderTopWidth = parseInt(
    scrollParentStyles.borderTopWidth,
    10
  )
  const scrollParentBorderBottomWidth = parseInt(
    scrollParentStyles.borderBottomWidth,
    10
  )
  const scrollParentTop = scrollParentRect.top + scrollParentBorderTopWidth
  const nodeRect = node.getBoundingClientRect()
  const nodeOffsetTop = nodeRect.top + scrollParent.scrollTop
  const nodeTop = nodeOffsetTop - scrollParentTop

  if (nodeTop < scrollParent.scrollTop) {
    // item is above the scrollable area
    scrollParent.scrollTop = nodeTop
    return
  }

  const nodeBottom = nodeTop + nodeRect.height
  const totalScrollBorderHeight =
    scrollParentBorderTopWidth + scrollParentBorderBottomWidth

  if (
    nodeBottom + totalScrollBorderHeight >
    scrollParent.scrollTop + scrollParentRect.height
  ) {
    // item is below the scrollable area
    scrollParent.scrollTop =
      nodeTop +
      nodeRect.height -
      scrollParentRect.height +
      scrollParentBorderTopWidth +
      scrollParentBorderBottomWidth
  }
  // item is within the scrollable area (do nothing)
}
