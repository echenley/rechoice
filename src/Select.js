// @flow

import * as React from 'react'
import cx from 'classnames'
import Menu from './Menu'
import contains from './util/contains'
import { addClickListener, removeClickListener } from './util/clickListener'
import keyCodes from './util/keyCodes'
import generateId from './util/generateId'
import getClasses from './util/getClasses'
import scrollIntoView from './util/scrollIntoView'
import {
  focusInput,
  focusPreviousItem,
  focusNextItem,
  updateInputValue,
} from './util/state'
import type {
  ClassMap,
  FilterItemsSingle,
  InputRenderer,
  Item,
  ItemRenderer,
  MenuRenderer,
  LoadingState,
} from './types'

type Props = {
  autoResizeInput: boolean,
  classes?: $Shape<ClassMap>,
  disabled: boolean,
  filterItems: FilterItemsSingle,
  getInputValue: (item: ?Item) => string,
  inputRef?: (ref: ?HTMLInputElement) => mixed,
  isOpen?: boolean,
  itemKey: string,
  items: Array<Item>,
  loadingIndicator?: React.Node,
  loadingState?: LoadingState,
  minimumInput: number,
  onBlur?: () => mixed,
  onChange: (item: ?Item) => mixed,
  onFocus?: () => mixed,
  onInputChange?: (inputValue: string) => mixed,
  placeholder?: string,
  renderInput?: InputRenderer,
  renderItem: ItemRenderer,
  renderMenu?: MenuRenderer,
  resetInputOnBlur?: boolean,
  rootRef?: (ref: ?HTMLElement) => mixed,
  value: ?Item,
}

type State = {
  filteredItems: Array<Item>,
  focusedItemIndex: ?number,
  inputValue: string,
  isFocused: boolean,
  isOpen: boolean,
}

class Select extends React.Component<Props, State> {
  static defaultProps = {
    autoResizeInput: false,
    resetInputOnBlur: true,
    disabled: false,
    filterItems: (items: Array<Item>) => items,
    minimumInput: 0,
  }

  constructor(props: Props, ...args: Array<Object>) {
    super(props, ...args)

    this.state = {
      filteredItems: props.items,
      focusedItemIndex: null,
      inputValue: props.value ? props.getInputValue(props.value) : '',
      isFocused: false,
      isOpen: false,
    }

    const id = generateId('rechoice')

    this.id = id
    this.controlsId = `${id}-listbox`
    this.classes = getClasses(props.classes)
  }

  blurTimer: ?TimeoutID
  focusedItemNode: ?HTMLElement
  inputNode: ?HTMLInputElement
  inputMirrorNode: ?HTMLElement
  menuNode: ?HTMLElement
  rootNode: ?HTMLElement

  id: string
  controlsId: string
  classes: ClassMap

  componentDidMount() {
    addClickListener(this.onDocumentClick)
  }

  componentWillReceiveProps(nextProps: Props) {
    const { filterItems, items } = nextProps

    if (items !== this.props.items) {
      this.setState(state => ({
        filteredItems: filterItems(items, state.inputValue, nextProps.value),
      }))
    }
  }

  componentDidUpdate() {
    const { inputNode, inputMirrorNode } = this
    const { autoResizeInput } = this.props

    if (!autoResizeInput || !inputNode || !inputMirrorNode) {
      return
    }

    inputNode.style.width = ''
    inputNode.style.width = `${inputMirrorNode.clientWidth}px`
  }

  componentWillUnmount() {
    removeClickListener(this.onDocumentClick)
    this.clearBlurTimer()
  }

  setFocusedItemRef = (ref: ?HTMLElement) => {
    this.focusedItemNode = ref
  }

  setInputMirrorRef = (ref: ?HTMLElement) => {
    this.inputMirrorNode = ref
  }

  setInputRef = (ref: ?HTMLInputElement) => {
    if (this.props.inputRef) {
      this.props.inputRef(ref)
    }

    this.inputNode = ref
  }

  setRootNode = (ref: ?HTMLElement) => {
    if (this.props.rootRef) {
      this.props.rootRef(ref)
    }

    this.rootNode = ref
  }

  setMenuRef = (ref: ?HTMLElement) => {
    this.menuNode = ref
  }

  close = () => {
    this.setState(state => {
      const {
        resetInputOnBlur,
        getInputValue,
        items,
        onInputChange,
        value,
      } = this.props

      const newState: $Shape<State> = {
        isOpen: false,
        isFocused: false,
      }

      if (resetInputOnBlur) {
        newState.inputValue = getInputValue(value)

        if (onInputChange && state.inputValue !== newState.inputValue) {
          onInputChange(newState.inputValue)
        }
      }

      if (value) {
        newState.filteredItems = items
      }

      return newState
    })
  }

  focusInput = () => {
    if (this.inputNode) {
      this.inputNode.focus()
    }
  }

  clearBlurTimer = () => {
    if (this.blurTimer) {
      clearTimeout(this.blurTimer)
      this.blurTimer = null
    }
  }

  onDocumentClick = (e: Event) => {
    const target = e.target
    const { rootNode, menuNode } = this

    if (target instanceof Node && rootNode && menuNode) {
      const isInRootNode = contains(rootNode, target)
      const isInMenu = contains(menuNode, target)
      if (isInMenu) {
        // stop menu click from blurring the input
        e.preventDefault()
      } else if (!isInRootNode && !isInMenu) {
        // close menu when clicking outside
        this.close()
      }
    }
  }

  onInputBlur = () => {
    this.clearBlurTimer()
    if (this.props.onBlur) {
      this.props.onBlur()
    }
    this.blurTimer = setTimeout(this.close, 10)
  }

  onInputFocus = () => {
    this.clearBlurTimer()
    if (this.props.onFocus) {
      this.props.onFocus()
    }
    this.setState(state => focusInput(state, this.props), this.scrollToSelected)
  }

  onKeyDown = (e: SyntheticKeyboardEvent<>) => {
    const {
      filteredItems,
      focusedItemIndex,
      inputValue,
      isFocused,
    } = this.state

    const {
      filterItems,
      loadingState,
      minimumInput,
      items,
      onChange,
      value,
    } = this.props

    if (e.keyCode === keyCodes.ENTER) {
      if (loadingState === 'loading') {
        e.preventDefault()
        return
      }

      if (
        typeof focusedItemIndex === 'number' &&
        isFocused &&
        this.getIsOpen()
      ) {
        this.onSelect(filteredItems[focusedItemIndex])
      } else if (inputValue.length >= minimumInput) {
        this.setState(state => ({
          isOpen: true,
          filteredItems: filterItems(items, state.inputValue, value),
        }))
      }
    } else if (e.keyCode === keyCodes.UP) {
      this.focusPreviousItem()
    } else if (e.keyCode === keyCodes.DOWN) {
      this.focusNextItem()
    } else if (
      value &&
      !inputValue &&
      (e.keyCode === keyCodes.BACKSPACE || e.keyCode === keyCodes.DELETE)
    ) {
      onChange(null)
    } else {
      return
    }

    e.preventDefault()
  }

  onHoverItem = (i: number) => {
    this.setState({
      focusedItemIndex: i,
    })
  }

  onSelect = (item: ?Item) => {
    if (!item) {
      return
    }

    const {
      onChange,
      onInputChange,
      getInputValue,
      filterItems,
      items,
    } = this.props

    onChange(item)

    const inputValue = getInputValue(item)

    if (onInputChange) {
      onInputChange(inputValue)
    }

    this.setState(
      {
        inputValue,
        isOpen: false,
        filteredItems: filterItems(items, inputValue, item),
      },
      this.focusInput
    )
  }

  onInputChange = (e: SyntheticInputEvent<HTMLInputElement>) => {
    const inputValue = e.target.value

    if (this.props.onInputChange) {
      this.props.onInputChange(inputValue)
    }

    this.setState(
      state => updateInputValue(state, this.props, inputValue),
      this.scrollToSelected
    )
  }

  focusPreviousItem() {
    this.setState(
      state => focusPreviousItem(state, this.props),
      this.scrollToSelected
    )
  }

  focusNextItem() {
    this.setState(
      state => focusNextItem(state, this.props),
      this.scrollToSelected
    )
  }

  scrollToSelected = () => {
    const { focusedItemNode, menuNode } = this

    if (!menuNode || !focusedItemNode) {
      return
    }

    scrollIntoView(focusedItemNode, menuNode)
  }

  isItemSelected = (item: Item) => {
    const { value, itemKey } = this.props
    return Boolean(value && value[itemKey] === item[itemKey])
  }

  getItemId = (index: number) => {
    return `${this.id}-item-${index}`
  }

  getIsOpen = () => {
    return typeof this.props.isOpen === 'undefined'
      ? this.state.isOpen
      : this.props.isOpen
  }

  render() {
    const {
      autoResizeInput,
      disabled,
      loadingIndicator,
      placeholder,
      itemKey,
      loadingState,
      renderInput,
      renderItem,
      renderMenu,
      value,
    } = this.props

    const {
      filteredItems,
      focusedItemIndex,
      inputValue,
      isFocused,
    } = this.state

    const selectClasses = cx(this.classes.root, {
      [this.classes.focused]: isFocused,
      [this.classes.disabled]: disabled,
    })

    const isOpen = this.getIsOpen()
    const focusedItemId =
      typeof focusedItemIndex === 'number'
        ? this.getItemId(focusedItemIndex)
        : null

    let inputMirror = null

    if (autoResizeInput) {
      inputMirror = (
        <span className={this.classes.inputMirror} ref={this.setInputMirrorRef}>
          {inputValue || placeholder}{' '}
        </span>
      )
    }

    let input = (
      <input
        aria-activedescendant={focusedItemId}
        aria-autocomplete="list"
        aria-owns={this.controlsId}
        autoComplete="off"
        className={this.classes.input}
        disabled={disabled}
        key="rechoice-input"
        onBlur={this.onInputBlur}
        onChange={this.onInputChange}
        onFocus={this.onInputFocus}
        onKeyDown={this.onKeyDown}
        placeholder={inputValue ? null : placeholder}
        ref={this.setInputRef}
        type="text"
        value={inputValue}
      />
    )

    if (renderInput) {
      input = renderInput({
        input,
        inputMirror,
        inputValue,
        value,
      })
    }

    return (
      <div className={selectClasses}>
        <div
          className={this.classes.inputWrapper}
          onClick={this.focusInput}
          ref={this.setRootNode}
          role="presentation"
        >
          {input}
        </div>
        <Menu
          classes={this.classes}
          controlsId={this.controlsId}
          focusedItemIndex={focusedItemIndex}
          focusedItemRef={this.setFocusedItemRef}
          getItemId={this.getItemId}
          inputValue={inputValue}
          isItemSelected={this.isItemSelected}
          isOpen={isOpen}
          itemKey={itemKey}
          items={filteredItems}
          loadingIndicator={loadingIndicator}
          menuRef={this.setMenuRef}
          onHoverItem={this.onHoverItem}
          onSelect={this.onSelect}
          renderItem={renderItem}
          renderMenu={renderMenu}
          state={loadingState}
        />
      </div>
    )
  }
}

export default Select
