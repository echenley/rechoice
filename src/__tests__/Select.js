/* eslint-disable max-nested-callbacks */

jest.mock('../util/clickListener')
jest.mock('../util/contains')
jest.mock('../util/scrollIntoView')

jest.useFakeTimers()

import * as React from 'react'
import { shallow } from 'enzyme'
import { addClickListener, removeClickListener } from '../util/clickListener'
import contains from '../util/contains'
import keyCodes from '../util/keyCodes'
import scrollIntoView from '../util/scrollIntoView'
import Select from '../Select'
import Menu from '../Menu'

contains.mockImplementation(() => false)

const value = { id: 3, label: 'three' }

// prettier-ignore
const items = [
  { id: 1, label: 'one' },
  { id: 2, label: 'two' },
  value,
  { id: 4, label: 'four' },
  { id: 5, label: 'five' },
]

const renderComponent = props => {
  return shallow(
    <Select
      getInputValue={item => item.label}
      itemKey="id"
      items={items}
      loadingIndicator={<div className="loading-indicator" />}
      onBlur={() => {}}
      onChange={() => {}}
      onFocus={() => {}}
      placeholder="Placeholderâ€¦"
      renderItem={() => {}}
      value={value}
      {...props}
    />
  )
}

describe('Select', () => {
  afterEach(() => {
    clearTimeout.mockClear()
    scrollIntoView.mockClear()
  })

  it('renders correctly', () => {
    expect(renderComponent().getElement()).toMatchSnapshot()
  })

  it('renders correctly when isOpen state to be overridden by props.isOpen', () => {
    const component = renderComponent({ isOpen: true })
    expect(component.find(Menu).prop('isOpen')).toBe(true)
  })

  it('adds "disabled" class and disables input when disabled', () => {
    const component = renderComponent({ disabled: true })
    expect(component.hasClass('disabled')).toBe(true)
    expect(component.find('input').prop('disabled')).toBe(true)
  })

  it('adds focused class when isFocused and calls onFocus props if provided', () => {
    const onFocus = jest.fn()
    const component = renderComponent({ onFocus })
    component.instance().onInputFocus()
    component.update()
    expect(component.hasClass('focused')).toBe(true)
    expect(onFocus).toHaveBeenCalled()
  })

  it('calls scrollIntoView when input is focused', () => {
    const component = renderComponent()
    const instance = component.instance()

    instance.menuNode = { id: 'menuNode' }
    instance.focusedItemNode = { id: 'focusedItemNode' }

    component.find('input').prop('onFocus')()

    expect(scrollIntoView).toHaveBeenCalledWith(
      instance.focusedItemNode,
      instance.menuNode
    )
  })

  describe('Methods', () => {
    describe('componentDidMount', () => {
      it('calls addClickListener', () => {
        const component = renderComponent()
        component.instance().componentDidMount()
        expect(addClickListener).toHaveBeenCalledWith(
          component.instance().onDocumentClick
        )
      })
    })

    describe('componentWillReceiveProps', () => {
      it('updates filteredItems if props.items changes', () => {
        const component = renderComponent()
        const newItems = [...items, { id: 6 }]
        component.setProps({
          items: newItems,
        })

        expect(component.state('filteredItems')).toEqual(newItems)
      })
    })

    describe('componentWillUnmount', () => {
      it('calls removeClickListener', () => {
        const component = renderComponent()
        const instance = component.instance()
        instance.clearBlurTimer = jest.fn()
        instance.componentWillUnmount()

        expect(removeClickListener).toHaveBeenCalledWith(
          instance.onDocumentClick
        )
        expect(instance.clearBlurTimer).toHaveBeenCalled()
      })
    })

    describe('close', () => {
      it('sets closed state', () => {
        const onInputChange = jest.fn()
        const component = renderComponent({ onInputChange })
        component.setState({
          inputValue: 'something',
          isOpen: true,
          isFocused: true,
        })

        component.instance().close()

        expect(component.state('isOpen')).toBe(false)
        expect(component.state('isFocused')).toBe(false)
        expect(component.state('inputValue')).toBe('three')
        expect(onInputChange).toHaveBeenCalledWith('three')
      })

      it('does not clear inputValue if resetInputOnBlur is false', () => {
        const component = renderComponent({
          resetInputOnBlur: false,
        })
        component.setState({
          inputValue: 'something',
          isOpen: true,
          isFocused: true,
        })

        component.instance().close()

        expect(component.state('isOpen')).toBe(false)
        expect(component.state('isFocused')).toBe(false)
        expect(component.state('inputValue')).toBe('something')
      })
    })

    describe('focusInput', () => {
      it('calls .focus() on input node', () => {
        const component = renderComponent()
        const instance = component.instance()
        instance.inputNode = { focus: jest.fn() }
        instance.focusInput()
        expect(instance.inputNode.focus).toHaveBeenCalled()
      })

      it('does not call .focus() if input node is null', () => {
        const component = renderComponent()
        const instance = component.instance()
        instance.inputNode = undefined
        expect(() => instance.focusInput()).not.toThrow()
      })
    })

    describe('clearBlurTimer', () => {
      it('calls clearTimeout on this.blurTimer', () => {
        const component = renderComponent()
        const instance = component.instance()
        instance.blurTimer = 111
        instance.clearBlurTimer()
        expect(clearTimeout).toHaveBeenCalledWith(111)
      })

      it('is a no-op if this.blurTimer is null', () => {
        const component = renderComponent()
        const instance = component.instance()
        instance.blurTimer = null
        instance.clearBlurTimer()
        expect(clearTimeout).not.toHaveBeenCalled()
      })
    })

    describe('onDocumentClick', () => {
      it('calls this.close if click target was not within the select or popover', () => {
        const component = renderComponent()
        const instance = component.instance()
        instance.close = jest.fn()
        instance.rootNode = {}
        instance.menuNode = {}
        instance.onDocumentClick({ target: document.createElement('div') })
        expect(instance.close).toHaveBeenCalled()
      })

      it('does not call this.close if click target was outside the select and popover', () => {
        contains.mockImplementationOnce(() => true)
        const component = renderComponent()
        const instance = component.instance()
        instance.close = jest.fn()
        instance.onDocumentClick({ target: {} })
        expect(instance.close).not.toHaveBeenCalled()
      })
    })

    describe('onHoverItem', () => {
      it('sets focusedItemIndex', () => {
        const component = renderComponent()
        component.find(Menu).prop('onHoverItem')(10)
        expect(component.state('focusedItemIndex')).toBe(10)
      })
    })

    describe('onInputBlur', () => {
      it('calls onBlur prop if provided', () => {
        const onBlur = jest.fn()
        const component = renderComponent({
          onBlur,
        })
        const instance = component.instance()
        instance.close = jest.fn()
        instance.clearBlurTimer = jest.fn()
        instance.onInputBlur()

        expect(onBlur).toHaveBeenCalled()
      })

      it('calls clearBlurTimer and close after a setTimeout', () => {
        const component = renderComponent()
        const instance = component.instance()
        instance.close = jest.fn()
        instance.clearBlurTimer = jest.fn()
        instance.onInputBlur()

        expect(instance.clearBlurTimer).toHaveBeenCalled()
        expect(instance.close).not.toHaveBeenCalled()

        jest.runAllTimers()

        expect(instance.close).toHaveBeenCalled()
        expect(Boolean(instance.blurTimer)).toBe(true)
      })
    })

    describe('onInputFocus', () => {
      it('calls clearBlurTimer and updates state', () => {
        const component = renderComponent()
        const instance = component.instance()
        instance.clearBlurTimer = jest.fn()
        component.find('input').prop('onFocus')()
        expect(instance.clearBlurTimer).toHaveBeenCalled()
        expect(component.state()).toMatchSnapshot()
      })
    })

    describe('onKeyDown', () => {
      it('ENTER opens the popover and updates filteredItems if popover is closed', () => {
        const component = renderComponent({
          filterItems: () => [],
        })
        const instance = component.instance()

        component.setState({ isOpen: false })

        const e = {
          keyCode: keyCodes.ENTER,
          preventDefault: jest.fn(),
        }

        instance.onKeyDown(e)

        expect(component.state('isOpen')).toBe(true)
        expect(component.state('filteredItems')).toEqual([])
        expect(e.preventDefault).toHaveBeenCalled()
      })

      it('ENTER calls this.onSelect if popover is open and focused', () => {
        const component = renderComponent()
        const instance = component.instance()

        component.setState({
          focusedItemIndex: 2,
          inputValue: 'test',
          isOpen: true,
          isFocused: true,
        })

        const e = {
          keyCode: keyCodes.ENTER,
          preventDefault: jest.fn(),
        }

        instance.onSelect = jest.fn()
        instance.onKeyDown(e)

        expect(instance.onSelect).toHaveBeenCalledWith(value)
        expect(e.preventDefault).toHaveBeenCalled()
      })

      it('ENTER does nothing when loading', () => {
        const component = renderComponent({
          loadingState: 'loading',
        })
        const instance = component.instance()

        component.setState({
          focusedItemIndex: 2,
          inputValue: 'test',
          isOpen: true,
          isFocused: true,
        })

        const e = {
          keyCode: keyCodes.ENTER,
          preventDefault: jest.fn(),
        }

        instance.onSelect = jest.fn()
        instance.onKeyDown(e)

        expect(instance.onSelect).not.toHaveBeenCalled()
        expect(e.preventDefault).toHaveBeenCalled()
      })

      it('UP focuses previous item and calls scrollIntoView', () => {
        const component = renderComponent()
        const instance = component.instance()

        const e = {
          keyCode: keyCodes.UP,
          preventDefault: jest.fn(),
        }

        instance.menuNode = { id: 'menuNode' }
        instance.focusedItemNode = { id: 'focusedItemNode' }
        instance.focusPreviousItem = jest.fn(instance.focusPreviousItem)
        instance.onKeyDown(e)

        expect(instance.focusPreviousItem).toHaveBeenCalledWith()
        expect(e.preventDefault).toHaveBeenCalled()
        expect(scrollIntoView).toHaveBeenCalledWith(
          instance.focusedItemNode,
          instance.menuNode
        )
      })

      it('DOWN focuses next item and calls scrollIntoView', () => {
        const component = renderComponent()
        const instance = component.instance()

        const e = {
          keyCode: keyCodes.DOWN,
          preventDefault: jest.fn(),
        }

        instance.menuNode = { id: 'menuNode' }
        instance.focusedItemNode = { id: 'focusedItemNode' }
        instance.focusNextItem = jest.fn(instance.focusNextItem)
        instance.onKeyDown(e)

        expect(instance.focusNextItem).toHaveBeenCalledWith()
        expect(e.preventDefault).toHaveBeenCalled()
        expect(scrollIntoView).toHaveBeenCalledWith(
          instance.focusedItemNode,
          instance.menuNode
        )
      })

      function clearInputTest() {
        const onChange = jest.fn()
        const component = renderComponent({ onChange })
        const instance = component.instance()
        component.setState({ inputValue: '' })

        const e = {
          keyCode: keyCodes.BACKSPACE,
          preventDefault: jest.fn(),
        }

        instance.onKeyDown(e)

        expect(onChange).toHaveBeenCalledWith(null)
        expect(e.preventDefault).toHaveBeenCalled()
      }

      it(
        'BACKSPACE clears current selection if inputValue is empty',
        clearInputTest
      )

      it(
        'DELETE clears current selection if inputValue is empty',
        clearInputTest
      )

      it('does not preventDefault for any other keys', () => {
        const component = renderComponent()
        const instance = component.instance()

        const e = {
          keyCode: 100,
          preventDefault: jest.fn(),
        }

        instance.onKeyDown(e)
        expect(e.preventDefault).not.toHaveBeenCalled()
      })
    })

    describe('onSelect', () => {
      it('does nothing if item is null', () => {
        const onChange = jest.fn()
        const component = renderComponent({ onChange })
        const instance = component.instance()
        instance.onSelect(null)
        expect(onChange).not.toHaveBeenCalled()
      })

      it('sets input to new value, closes popover, and resets filteredItems', () => {
        const onChange = jest.fn()
        const onInputChange = jest.fn()
        const items = [{ id: 1 }, { id: 2 }, value, { id: 4 }]
        const component = renderComponent({
          onChange,
          onInputChange,
          items,
        })
        component.setState({
          filteredItems: [value],
          inputValue: 'test',
          isOpen: true,
        })
        const instance = component.instance()
        instance.focusInput = jest.fn()
        instance.onSelect(value)

        expect(onChange).toHaveBeenCalledWith(value)
        expect(instance.focusInput).toHaveBeenCalled()
        expect(component.state('inputValue')).toBe('three')
        expect(component.state('isOpen')).toBe(false)
        expect(component.state('filteredItems')).toBe(items)
        expect(onInputChange).toHaveBeenCalledWith('three')
      })
    })

    describe('onInputChange', () => {
      it('updates state with new input value', () => {
        const onInputChange = jest.fn()
        const component = renderComponent({
          onInputChange,
        })
        const instance = component.instance()
        instance.onInputChange({
          target: {
            value: 'test',
          },
        })

        expect(onInputChange).toHaveBeenCalledWith('test')
        expect(component.state()).toMatchSnapshot()
      })
    })

    describe('isItemSelected', () => {
      it('returns true if item matches props.value', () => {
        const component = renderComponent()
        expect(component.instance().isItemSelected({ id: 3 })).toBe(true)
      })

      it('returns false if item does not match props.value', () => {
        const component = renderComponent()
        expect(component.instance().isItemSelected({ id: 2 })).toBe(false)
      })
    })
  })
})
