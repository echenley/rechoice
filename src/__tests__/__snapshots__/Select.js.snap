// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Methods onInputChange updates state with new input value 1`] = `
Object {
  "filteredItems": Array [
    Object {
      "id": 1,
      "label": "one",
    },
    Object {
      "id": 2,
      "label": "two",
    },
    Object {
      "id": 3,
      "label": "three",
    },
    Object {
      "id": 4,
      "label": "four",
    },
    Object {
      "id": 5,
      "label": "five",
    },
  ],
  "focusedItemIndex": 2,
  "inputValue": "test",
  "isFocused": false,
  "isOpen": true,
}
`;

exports[`Select Methods onInputFocus calls clearBlurTimer and updates state 1`] = `
Object {
  "filteredItems": Array [
    Object {
      "id": 1,
      "label": "one",
    },
    Object {
      "id": 2,
      "label": "two",
    },
    Object {
      "id": 3,
      "label": "three",
    },
    Object {
      "id": 4,
      "label": "four",
    },
    Object {
      "id": 5,
      "label": "five",
    },
  ],
  "focusedItemIndex": 2,
  "inputValue": "three",
  "isFocused": true,
  "isOpen": true,
}
`;

exports[`Select renders correctly 1`] = `
<div
  className="rechoice"
>
  <div
    className="rechoice-input-wrapper"
    onClick={[Function]}
    role="presentation"
  >
    <input
      aria-activedescendant={null}
      aria-autocomplete="list"
      aria-owns="rechoice-0-listbox"
      autoComplete="off"
      className="rechoice-input"
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder={null}
      type="text"
      value="three"
    />
  </div>
  <Menu
    classes={
      Object {
        "disabled": "disabled",
        "focused": "focused",
        "input": "rechoice-input",
        "inputMirror": "rechoice-input-mirror",
        "inputWrapper": "rechoice-input-wrapper",
        "item": "rechoice-item",
        "menu": "rechoice-menu",
        "root": "rechoice",
        "selected": "selected",
        "tags": "rechoice-tags",
        "value": "rechoice-value",
      }
    }
    controlsId="rechoice-0-listbox"
    focusedItemIndex={null}
    focusedItemRef={[Function]}
    getItemId={[Function]}
    inputValue="three"
    isItemSelected={[Function]}
    isOpen={false}
    itemKey="id"
    items={
      Array [
        Object {
          "id": 1,
          "label": "one",
        },
        Object {
          "id": 2,
          "label": "two",
        },
        Object {
          "id": 3,
          "label": "three",
        },
        Object {
          "id": 4,
          "label": "four",
        },
        Object {
          "id": 5,
          "label": "five",
        },
      ]
    }
    loadingIndicator={
      <div
        className="loading-indicator"
      />
    }
    menuRef={[Function]}
    onHoverItem={[Function]}
    onSelect={[Function]}
    renderItem={[Function]}
    renderMenu={undefined}
    state={undefined}
  />
</div>
`;
